databases/jdp/database_test.go

```diff
@@ -0,0 +1,121 @@
+package jdp
+
+import (
+	"github.com/stretchr/testify/assert"
+	"os"
+	"testing"
+
+	"dts/config"
+	"dts/core"
+)
+
+const jdpConfig string = `
+databases:
+  jdp:
+    name: JGI Data Portal
+    organization: Joint Genome Institue
+    url: https://files.jgi.doe.gov
+    endpoint: globus-jdp
+    auth:
+      client_id: ${JGI_CLIENT_ID}
+      client_secret: ${JGI_CLIENT_SECRET}
+endpoints:
+  globus-jdp:
+    name: Globus NERSC DTN
+    id: ${DTS_GLOBUS_TEST_ENDPOINT}
+    provider: globus
+    auth:
+      client_id: ${DTS_GLOBUS_CLIENT_ID}
+      client_secret: ${DTS_GLOBUS_CLIENT_SECRET}
+`
+
+// this function gets called at the begіnning of a test session
+func setup() {
+	config.Init([]byte(jdpConfig))
+}
+
+// this function gets called after all tests have been run
+func breakdown() {
+}
+
+func TestNewDatabase(t *testing.T) {
+	assert := assert.New(t) // binds assert to t
+	orcid := os.Getenv("DTS_KBASE_TEST_ORCID")
+	jdpDb, err := NewDatabase(orcid)
+	assert.NotNil(jdpDb, "JDP database not created")
+	assert.Nil(err, "JDP database creation encountered an error")
+}
+
+func TestNewDatabaseWithoutOrcid(t *testing.T) {
+	assert := assert.New(t) // binds assert to t
+	jdpDb, err := NewDatabase("")
+	assert.Nil(jdpDb, "Invalid JDP database somehow created")
+	assert.NotNil(err, "JDP database creation without ORCID encountered no error")
+}
+
+func TestNewDatabaseWithoutJDPSharedSecret(t *testing.T) {
+	assert := assert.New(t) // binds assert to t
+	orcid := os.Getenv("DTS_KBASE_TEST_ORCID")
+	jdpSecret := os.Getenv("DTS_JDP_SECRET")
+	os.Unsetenv("DTS_JDP_SECRET")
+	jdpDb, err := NewDatabase(orcid)
+	os.Setenv("DTS_JDP_SECRET", jdpSecret)
```

Comment:
will the `DTS_JDP_SECRET` be set again if `NewDatabase` throws an error / fails in some way? i.e. will `DTS_JDP_SECRET` always get re-set?

Reply:
Yes, it always gets reset, because we reset it before acting on any error conditions. In departing from exception-based error flows, Go (like C) makes one respond to error codes instead. A lot of folks don't like this, but in my experience, there's no way to do error handling that is simultaneously flexible (handle your own errors the way you think they should be handled, when you think they should be handled!) and seamless (toss an exception into the ether and forget about it, and be prepared to catch these things downstream!).

Reply:
Nice. Setting/unsetting env vars in python tests reliably requires a lot more fiddling so it's nice to see how much easier it is here!